engine/audiomixer.cpp

```diff
@@ -0,0 +1,167 @@
+#include "audiomixer.h"
+
+#include "config.h"
+#include "logger.h"
+#include "pong.h"
+#include "sound.h"
+
+#include <glm/glm.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <chrono>
+#include <iostream>
+
+namespace pong
+{
+
+static constexpr int SAMPLE_RATE = 44100;
+static constexpr int FRAMES_PER_BUFFER = 256;
+static constexpr glm::vec3 LEFT_LISTENER_POSITION = glm::vec3(-1125.0f, 0.0f, 0.0f);
+static constexpr glm::vec3 RIGHT_LISTENER_POSITION = glm::vec3(1125.0f, 0.0f, 0.0f);
+static constexpr float MIN_DISTANCE = 250.0f;
+
+int AudioCallbackWrapper(const void* inputBuffer, void* outputBuffer,
+                  unsigned long framesPerBuffer,
+                  const PaStreamCallbackTimeInfo* timeInfo,
+                  PaStreamCallbackFlags statusFlags,
+                  void* userData)
+{
+    return Pong::GetInstance().GetAudioMixer().AudioCallback(inputBuffer, outputBuffer, framesPerBuffer, timeInfo, statusFlags, userData);
+}
+
+bool AudioMixer::Init()
+{
+    mVolume = Config::GetValue("volume", 0.5f);
+    LogInfo("Volume: {}", mVolume);
+
+    mSpatialAudioEnabled = Config::GetValue("spatial_audio", false);
+    LogInfo("Spatial Audio: {}", mSpatialAudioEnabled ? "Enabled" : "Disabled");
+
+    PaError err = Pa_Initialize();
+    if (err != paNoError) {
+        LogError("PortAudio initialization failed: {}", Pa_GetErrorText(err));
+    }
+
+    PaStreamParameters outputParameters;
+    outputParameters.device = Pa_GetDefaultOutputDevice();
+    outputParameters.channelCount = 2; // Stereo
+    outputParameters.sampleFormat = paFloat32;
+    outputParameters.suggestedLatency = Pa_GetDeviceInfo(outputParameters.device)->defaultLowOutputLatency;
+    outputParameters.hostApiSpecificStreamInfo = nullptr;
+
+    err = Pa_OpenStream(&mStream, nullptr, &outputParameters, SAMPLE_RATE, FRAMES_PER_BUFFER, paClipOff, AudioCallbackWrapper, nullptr);
+    if (err != paNoError) {
+        LogError("PortAudio stream opening failed: {}", Pa_GetErrorText(err));
+        Pa_Terminate();
+        return false;
```

Comment:
could be an opportunity to just assert and die here. Seems like not opening the device is a critical failure and everything should be aborted immediately, same with below